File: Android/PlayerProj/animplayer/src/main/java/com/tencent/qgame/animplayer/AnimConfig.kt
/*
 * Tencent is pleased to support the open source community by making vap available.
 *
 * Copyright (C) 2020 THL A29 Limited, a Tencent company.  All rights reserved.
 *
 * Licensed under the MIT License (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 *
 * http://opensource.org/licenses/MIT
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is
 * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tencent.qgame.animplayer

import android.graphics.Bitmap
import android.graphics.MaskFilter
import com.tencent.qgame.animplayer.mask.MaskConfig
import com.tencent.qgame.animplayer.util.ALog
import org.json.JSONException
import org.json.JSONObject

/**
 * vapc里读取出来的基础配置
 */
class AnimConfig {

    companion object {
        private const val TAG = "${Constant.TAG}.AnimConfig"
    }

    val version = 2 // 不同版本号不兼容
    var totalFrames = 0 // 总帧数
    var width = 0 // 需要显示视频的真实宽高
    var height = 0
    var videoWidth = 0 // 视频实际宽高
    var videoHeight = 0
    var orien = Constant.ORIEN_DEFAULT // 0-兼容模式 1-竖屏 2-横屏
    var fps = 0
    var isMix = false // 是否为融合动画
    var alphaPointRect = PointRect(0, 0 ,0 ,0) // alpha区域
    var rgbPointRect = PointRect(0, 0, 0, 0) // rgb区域
    var isDefaultConfig = false // 没有vapc配置时默认逻辑
    var defaultVideoMode = Constant.VIDEO_MODE_SPLIT_HORIZONTAL

    var maskConfig: MaskConfig ?= null
    var jsonConfig: JSONObject? = null


    /**
     * @return 解析是否成功，失败按默认配置走
     */
    fun parse(json: JSONObject): Boolean {
        return try {
            json.getJSONObject("info").apply {
                val v = getInt("v")
                if (version != v) {
                    ALog.e(TAG, "current version=$version target=$v")
                    return false
                }
                totalFrames = getInt("f")
                width = getInt("w")
                height = getInt("h")
                videoWidth = getInt("videoW")
                videoHeight = getInt("videoH")
                orien = getInt("orien")
                fps = getInt("fps")
                isMix = getInt("isVapx") == 1
                val a = getJSONArray("aFrame") ?: return false
                alphaPointRect = PointRect(a.getInt(0), a.getInt(1), a.getInt(2), a.getInt(3))
                val c = getJSONArray("rgbFrame") ?: return false
                rgbPointRect = PointRect(c.getInt(0), c.getInt(1), c.getInt(2), c.getInt(3))
            }
            true
        } catch (e : JSONException) {
            ALog.e(TAG, "json parse fail $e", e)
            false
        }
    }

    override fun toString(): String {
        return "AnimConfig(version=$version, totalFrames=$totalFrames, width=$width, height=$height, videoWidth=$videoWidth, videoHeight=$videoHeight, orien=$orien, fps=$fps, isMix=$isMix, alphaPointRect=$alphaPointRect, rgbPointRect=$rgbPointRect, isDefaultConfig=$isDefaultConfig)"
    }


}


data class PointRect(val x: Int, val y: Int, val w: Int, val h: Int)
data class RefVec2(val w: Int, val h: Int) //参考宽&高

File: ohos/vap_module/src/main/cpp/render/anim_config.cpp
/*
 * Copyright (C) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <fstream>
#include "common_const.h"
#include "anim_config.h"

using namespace CommonConst;

AnimConfig::~AnimConfig()
{
    if (srcMapPtr) {
        srcMapPtr->srcSMap.clear();
    }
    if (frameAllPtr) {
        frameAllPtr->frameAll.clear();
    }
};

void AnimConfig::ParseBoxHead(std::vector<char> &boxHead, BoxHead &head)
{
    int32_t length = ZERO;
    length |= (boxHead[ZERO] & ONE_EIGHT_HEX) << THREE_TIME_EIGHT;
    length |= (boxHead[ZERO] & ONE_EIGHT_HEX) << TWO_TIME_EIGHT;
    length |= (boxHead[TWO] & ONE_EIGHT_HEX) << EIGHT;
    length |= (boxHead[THREE] & ONE_EIGHT_HEX);
    head.length = length;
    head.type.assign(&boxHead[FOUR], &boxHead[FOUR] + FOUR);
}

int AnimConfig::GetJson(std::string &jsonStr, std::string &uri)
{
    std::ifstream file(uri, std::ios::binary);
    int ret = NEGATIVE_ONE;
    if (!file.is_open()) {
        LOGE("json Failed to open file");
        return ret;
    }
    std::vector<char> buffer(EIGHT);
    while (file.read(buffer.data(), EIGHT)) {
        if (file.eof()) {
            LOGD("json end of file");
            break;
        }
        BoxHead boxHead;
        ParseBoxHead(buffer, boxHead);

        if (boxHead.type.compare("vapc") == ZERO) { // memcmp(boxHead.type.data(), "vapc", 4) == 0
            LOGD("json length: %{public}d", boxHead.length);
            std::vector<char> cont(boxHead.length);
            file.read(cont.data(), boxHead.length);
            jsonStr.assign(cont.begin(), cont.end());
            ret = 0;
            break;
        }
        if (isH265_) {
            file.seekg(ONE - EIGHT, std::ios::cur);
        }
    }
    file.close();
    return ret;
}

bool AnimConfig::ParseJson(std::string uri, bool isNeedParseFrame)
{
    LOGD("Enter ParseJson");
    if (isInit) {
        LOGD("Json data has been initialized");
        return true;
    }
    std::string strJson;
    json jsonObj;
    if (GetJson(strJson, uri) == NEGATIVE_ONE) {
        LOGE("AnimConfig::String2Json failed");
        return false;
    }
    if (strJson.empty()) {
        LOGE("AnimConfig not get json");
        return false;
    }
    jsonObj = json::parse(strJson);
    if (jsonObj.find("info") == jsonObj.end()) {
        LOGE("jsonObj: not find info");
        return false;
    }
    const auto &info = jsonObj.at("info");

    this->width = info.at("w").get<int>();
    this->height = info.at("h").get<int>();
    this->totalFrames = info.at("f").get<int>();
    this->videoWidth = info.at("videoW").get<int>();
    this->videoHeight = info.at("videoH").get<int>();
    this->orien = info.at("orien").get<Orien>();
    this->fps = info.at("fps").get<int>();
    this->isMix = info.at("isVapx").get<int>() == ONE;
    const auto& a = info.at("aFrame");
    this->alphaPointRect = std::move(PointRect(a.at(ZERO).get<int>(), a.at(ONE).get<int>(),
        a.at(TWO).get<int>(), a.at(THREE).get<int>()));
    const auto& c = info.at("rgbFrame");
    this->rgbPointRect = std::move(PointRect(c.at(ZERO).get<int>(), c.at(ONE).get<int>(),
        c.at(TWO).get<int>(), c.at(THREE).get<int>()));

    if (jsonObj.find("src") != jsonObj.end()) {
        srcMapPtr = std::make_shared<SrcMap>(jsonObj);
    } else {
        LOGW("json: not find src");
    }

    if (isNeedParseFrame && jsonObj.find("frame") != jsonObj.end()) {
        frameAllPtr = std::make_shared<FrameAll>(jsonObj);
    } else {
        LOGW("json: not find frame");
    }
    
    isInit = true;
    LOGD("end ParseJson");
    return true;
}

void AnimConfig::DefaultConfig(int32_t width, int32_t height)
{
    this->videoWidth = width;
    this->videoHeight = height;
    
    switch (defaultVideoMode) {
        case VIDEO_MODE_SPLIT_VERTICAL:
            this->width = width;
            this->height = height / TWO;
            this->alphaPointRect   = {ZERO, ZERO, this->width, this->height};
            this->rgbPointRect = {ZERO, this->height, this->width, this->height};
            break;
        case VIDEO_MODE_SPLIT_HORIZONTAL_REVERSE:
            this->width = width / TWO;
            this->height = height;
            this->rgbPointRect = {ZERO, ZERO, this->width, this->height};
            this->alphaPointRect = {this->width, ZERO, this->width, this->height};
            break;
        case VIDEO_MODE_SPLIT_VERTICAL_REVERSE:
            this->width = width;
            this->height = height / TWO;
            this->rgbPointRect = {ZERO, ZERO, this->width, this->height};
            this->alphaPointRect = {ZERO, this->height, this->width, this->height};
            break;
        default:
            this->width = width / TWO;
            this->height = height;
            this->alphaPointRect   = {ZERO, ZERO, this->width, this->height};
            this->rgbPointRect = {this->width, ZERO, this->width, this->height};
            break;
    }
}


void AnimConfig::GetSrc(std::map<std::string, Src> &src)
{
    if (!srcMapPtr) {
        LOGW("this file not have src prop");
    } else {
        src = srcMapPtr->srcSMap;
    }
}
